### PR TITLE
Use -XX:ArchiveClassesAtExit for AppCDS creation in Java 17+

### PR DESCRIPTION
**Pull request by geoand**
Not only does this simplify the process of generating the AppCDS,
but for some reason it also leads to an archive that is both smaller
and more useful for the JVM (meaning that it reduces startup time
more).

For example using the `rest-json-quickstart` on my machine I got the following:

_Without AppCDS_: **~450ms**
_With old AppCDS_: **~400ms**
_With new AppCDS_: **~340ms**

### CODE REVIEW

**Comment by geoand** [comment]
@loicmathieu you'll probably appreciate this one

**Comment by loicmathieu** [comment]
@geoand this is indead interesting! It should not make any difference, on the same Java version, to make it one way or the other, as far as I know the new option is just for simplicity sake.

I need to give it a second try as I tested it a few months ago but didn't deploy it in production because on my CI a different image is used to build the apps than the one that runs it, and sometimes there is version divergence (not the same JVM build) and AppCDS require the exact same JVM version not just the same major version.

Anyway, great PR.

File: core/runtime/src/main/java/io/quarkus/runtime/shutdown/ShutdownRecorder.java [hunk 1]
```diff
@@ -30,6 +30,9 @@ public void setListeners(List<ShutdownListener> listeners) {
     }
 
     public static void runShutdown() {
+        if (shutdownListeners == null) {
+            return;
+        }
```
Comment (loicmathieu) [inline core/runtime/src/main/java/io/quarkus/runtime/shutdown/ShutdownRecorder.java:35]: Best to move it under the debug log

**Review by loicmathieu — APPROVED** [review]
LGTM

File: core/runtime/src/main/java/io/quarkus/runtime/shutdown/ShutdownRecorder.java (hunk 1 above)
Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/runtime/shutdown/ShutdownRecorder.java:35]: I actually think this one makes a little more sense

**Comment by geoand** [comment]
@Sanne  @maxandersen this makes our AppCDS story even better.

There are still things we can (like @Sanne's idea of having extensions contribute to the warmup - although in a test I did with Jackson it did not make any difference), but I feel that with this, we are at last in a pretty good place.

**Comment by Sanne** [comment]
@geoand very nice!

Regarding the idea of having extensions "contribute" - how did you model that experiment, at high level?

I was expecting the extension could either trigger some operations (needs some ad-hoc knowledge to make some choices which have no durable side-effects and actually do triggers some kind of useful warmup), or list some key classed for forceful initialization. More likely a combination. Is that similar to your experiment?

**Comment by loicmathieu** [comment]
Shouldn't some of the new [PreloadClassesBuildStep](https://github.com/quarkusio/quarkus/blob/431666d385297600d790a7cb14cb701f1b93c6a9/core/deployment/src/main/java/io/quarkus/deployment/steps/PreloadClassesBuildStep.java) and it's processor be reused for this ?

It was designed for CRaC but it's more or less the same principle: having a way to preload classes as some JVM initialization class to warmup the JVM.

**Comment by geoand** [comment]
Yes, there are definitely possible synergies there. I just wanted to throw together something quick and dirty which is why I tried ^

**Comment by loicmathieu** [comment]
Yeah, this can be done as a followup to share init code for CRaC, AppCDS and SnapStart

**Comment by Sanne** [comment]
I like were this is going :) 

N.B. Not sure if you tested, but when doing pre-loading of classes we need to also to make sure it's happening in the same classloader as the app would normally load them with.

**Comment by geoand** [comment]
> N.B. Not sure if you tested, but when doing pre-loading of classes we need to also to make sure it's happening in the same classloader as the app would normally load them with.

Yup, I've been bitten by that one in the past when looking into AppCDs :)

**Comment by Sanne** [comment]
p.s. the JVM has some flags (don't remember them on top of my head) to log class loading events as they happen. Might be useful to validate some approaches.

**Comment by loicmathieu** [comment]
`-Xlog:class,cds`  should display everything !

`cds` only will print issues like JVM mismatch for ex.

**Pull request merged**